Move MSBuild projects out of the native build scripts

jkoritzinsky opened the pull request:
### Summary

Significantly simplify our native build scripts by moving all of the managed builds up to the Arcade build infrastructure and splitting them into subsets to help save developer time on their inner builds.

Additionally, since it isn't easily possible to make Arcade build a project as self-contained publish, this PR also changes the crossgen2 build to build framework-dependent, have the Helix SDK pull down a .NET CLI for the test runs, and update the test run scripts to resolve the local dotnet CLI or the one in the path based on environment variables.

### Subsets

The coreclr build is split into these subsets:

- corelib
  - Contains System.Private.CoreLib
- runtime
  - The native build as well as crossgenning of System.Private.CoreLib
- linuxdac (Windows-only)
  - The cross-OS Windows->Linux DAC
- alpinedac (Windows-only)
  - The cross-OS Windows->Alpine Linux DAC
- nativecorelib
  - Crossgens System.Private.Corelib
- tools
  - The managed tools such as R2RDump and Crossgen2
- packages
  - The coreclr packages

### Local build improvements

The managed eventing headers are generated during the System.Private.CoreLib build by discovering python and running the script during the build in MSBuild.

The ucrt copy and enforce-pgo steps run as part of the native build instead of the managed build, which makes significantly more sense logically.

Due to keeping all of the managed builds within a single MSBuild invocation instead of nested invocations, this change saves around a minute in local build times.

### Workflow changes

CoreCLR developers will have to use the root `build.cmd/sh` scripts to build any of the managed components. The `src/coreclr/build-rutnime.cmd/sh` scripts only support building the native coreclr CMake project. The `src/coreclr/crossgen-corelib.cmd/sh` scripts support running crossgen on System.Private.CoreLib. The old `src/coreclr/build.cmd/sh` scripts are kept so as to give developers a transition period from the scripts to the new scripts.

#### Workflow examples

I've included some workflow examples below.

Build CoreCLR, System.Private.CoreLib, crossgen System.Private.CoreLib, managed tools, CoreCLR packages in Checked configuration for x86:

`D:\source\runtime>./build.cmd -subsetCategory coreclr -c checked -arch x86`

Build CoreCLR, System.Private.CoreLib, crossgen System.Private.CoreLib, skip managed tools, skip coreclr packages:

`D:\source\runtime>./build.cmd -subsetcategory coreclr -subset runtime-corelib-nativecorelib`

Build CoreCLR and managed tools only:

`D:\source\runtime>./build.cmd -subsetcategory coreclr -subset runtime-tools`

Build CoreCLR, System.Private.CoreLib, crossgen System.Private.CoreLib, managed tools, CoreCLR packages in debug and libraries in release (already works today):

`D:\source\runtime> ./build.cmd -subsetcategory coreclr-libraries -c Release -runtimeConfiguration debug`

### Future improvements

This PR will enable us to build System.Private.CoreLib in a job before running the coreclr or libraries builds. As a result, this moves us closer to being able to build coreclr and libraries in parallel in CI. Additionally, this does most of the work to Arcadify coreclr's build, so we can focus more on what our next steps and "final" build script design will be.

jkoritzinsky committed:
Move coreclr managed tools that don't need to be published and System.Private.CoreLib into their own subsets in Subsets.props and drive their builds via Arcade instead of the coreclr scripts.

jkoritzinsky committed:
Remove building System.Private.CoreLib from build scripts. Eventing sources TBD.

jkoritzinsky committed:
Build packages via an Arcade subset instead of within build.cmd/build.sh

jkoritzinsky committed:
Remove packages and test build from build.cmd. Clean up options on build.cmd and build.sh

jkoritzinsky committed:
Generate eventing sources for System.Private.CoreLib within MSBuild instead of in the build.cmd/sh scripts. Clean up the build.sh usage documentation (lots of copy-paste errors).

jkoritzinsky committed:
Remove check-definitions.py script and supporting goo.

jkoritzinsky committed:
Fix eventing generation.

jkoritzinsky committed:
Wire up  TargetArchitecture and OSGroup to coreclr's BuildArch and BuildOS to enable building System.Private.CoreLib for the correct output given the script parameters.

jkoritzinsky committed:
Stop passing down now non-existent skiptests parameter.

jkoritzinsky committed:
Fix coreclr .nuget rid parsing to account for linux-musl correctly.

jkoritzinsky committed:
Update build scripts to use Arcade to build packages.

jkoritzinsky committed:
Merge branch 'master' of https://github.com/dotnet/runtime into hoist-S.P.CL

jkoritzinsky committed:
Convert from Properties to AdditionalProperties.

jkoritzinsky committed:
Build crossgen as part of Arcade (self-contained publish TBD).

jkoritzinsky committed:
Pull down .NET runtime CLI for helix runs with crossgen2.

jkoritzinsky committed:
Output crossgen2 to the BinDir/crossgen when built from the project file. Enable the test shell scripts to either resolve the dotnet CLI from the path or from relative to Core_Root if the __DotNet variable is undefined.

jkoritzinsky committed:
Generate distro rid during windows build.

jkoritzinsky committed:
Various fixes to get files binplaced correctly and builds to work on Linux.

jkoritzinsky committed:
Merge branch 'hoist-S.P.CL' of https://github.com/jkoritzinsky/runtime into hoist-S.P.CL

jkoritzinsky committed:
Merge branch 'master' of https://github.com/dotnet/runtime into hoist-S.P.CL

jkoritzinsky committed:
Fix search for dotnet in repo on bash.

jkoritzinsky committed:
Remove __BuildTest variable and fix up the script.

jkoritzinsky committed:
Fix steps in build-job to build S.P.CL before the runtime and build managed tools after the runtime.

AaronRobinsonMSFT commented:
>To help make this more visible, I'm thinking about renaming the build.cmd/sh scripts to be build-native.cmd/sh and splitting out the crossgen script to be more explicit.

This renaming is a good idea that I strongly suggest. Instead of completely removing the `build.cmd/sh` scripts I would instead leave them intact for some amount of time (2 - 3 weeks) and have them print out the new workflow (i.e. point to a readme/URL/PR). The breadcrumb of what to do next would be very helpful.

am11 commented on eng/build.sh:
```diff
@@ -58,13 +58,104 @@ usage()
+
+# Use uname to determine what the OS is.
+OSName=$(uname -s)
+case "$OSName" in
```
Could we factor out CpuName and OSName blocks from https://github.com/dotnet/runtime/blob/bc5e7e36cefd1b53b654419cfd2bffc6b65dd403/eng/native/build-commons.sh#L240 in a separate file (something like `eng/native/initOSAndArch.sh`) and reuse? It used to be in four places and recently converged in a common script.

jkoritzinsky committed:
Fix casing typo.

jkoritzinsky committed:
Refactor detecting system OS and Architecture.

jkoritzinsky committed:
Fix wasm rid calculation.

jkoritzinsky committed:
Run crossgen2 on corerun so we run it on the runtime we're going to ship it with.

jkoritzinsky committed:
Assemble a live self-contained publish in the crossgen2 package.

jkoritzinsky committed:
Fix init-os-and-arch resolution.

jkoritzinsky committed:
Ensure that the runtime files we package in crossgen2 are the crossgened ones.

jkoritzinsky committed:
Remove extraneous /bl

jkoritzinsky committed:
Build R2RDump to a subfolder instead of the product dir.

jkoritzinsky committed:
Merge branch 'hoist-S.P.CL' of E:/home/jekoritz/runtime into hoist-S.P.CL

jkoritzinsky committed:
Merge branch 'master' of https://github.com/dotnet/runtime into hoist-S.P.CL

jkoritzinsky committed:
Seperate out the crosssgening corelib step into a separate script.

jkoritzinsky committed:
Fix permissions on crossgen-corelib.sh

jkoritzinsky committed:
Merge branch 'master' of https://github.com/dotnet/runtime into hoist-S.P.CL

jkoritzinsky committed:
Rename build.cmd/sh to build-runtime.cmd/sh

jkoritzinsky committed:
Fix parameters in crossgen2determinism test.

jkoritzinsky committed:
Merge branch 'master' of https://github.com/dotnet/runtime into hoist-S.P.CL

jkoritzinsky committed:
Run crossgen on System.Private.CoreLib in CI.

jkoritzinsky committed:
Restore original build.cmd/sh scripts with a deprecation warning and update instructions.

jkoritzinsky committed:
Initialize developer prompt in crossgen-corelib.cmd script so we can find DIA on CI machines.

jkoritzinsky committed:
Merge branch 'hoist-S.P.CL' of https://github.com/jkoritzinsky/runtime into hoist-S.P.CL

jkoritzinsky committed:
Make sure vs tools are enabled so we can resolve DiaSymReader.

jkoritzinsky committed:
Call build-native instead of build.

jkoritzinsky committed:
Merge branch 'master' of https://github.com/dotnet/runtime into hoist-S.P.CL

Restore build.proj for deprecated build script.

jkoritzinsky committed:
Fix build-native -> build-runtime.

jkoritzinsky committed:
Fix permissions.

jkoritzinsky commented:
Failures are https://github.com/dotnet/runtime/issues/32126

jkoritzinsky commented:
@fadimounir can you test out a crossgen2 package from this PR to make sure it lays everything out correctly for the SDK to consume it?

jaredpar commented on src/coreclr/build.cmd:
```diff
@@ -6,6 +6,7 @@ setlocal EnableDelayedExpansion EnableExtensions
 set "__MsgPrefix=BUILD: "
 
 echo %__MsgPrefix%Starting Build at %TIME%
+echo %__MsgPrefix%WARNING: This build script is deprecated and will be deleted soon. Use the root build script to build CoreCLR. If you want to build the CoreCLR runtime without using MSBuild, use the build-native.cmd script.
```
Please file an issue that tracks the deletion of the sub-directory build scripts and add a link above this line. This way we broadcast our intent here and give contributors a place to discuss the issue.

jaredpar commented on src/coreclr/build.sh:
```diff
@@ -386,6 +386,7 @@ handle_arguments_local() {
 }
 
 echo "Commencing CoreCLR Repo build"
+echo "WARNING: This build script is deprecated and will be deleted soon. Use the root build script to build CoreCLR. If you want to build the CoreCLR runtime without using MSBuild, use the build-native.sh script."
```
Please file an issue that tracks the deletion of the sub-directory build scripts and add a link above this line. This way we broadcast our intent here and give contributors a place to discuss the issue.

jaredpar commented on src/coreclr/build-runtime.sh:
```diff
@@ -0,0 +1,273 @@
+#!/usr/bin/env bash
```
Is this script meant for direct developer usage as say `build.cmd` formerly was?

jkoritzinsky replied:
Yes. For people who want to avoid using MSBuild as much as possible or for people who are building on platforms where we don't have a working MSBuild, they can call this script directly. It should have the same user experience as the old build.cmd/sh scripts except it only builds the CoreCLR runtime and cross components.

jaredpar commented on src/coreclr/clr.featuredefines.props:
```diff
@@ -76,4 +76,13 @@
         <DefineConstants Condition="'$(TargetsOSX)' == 'true'">$(DefineConstants);TARGET_OSX</DefineConstants>
     </PropertyGroup>
+
+  <!-- PYTHON will not be defined when building in VS-->
```
Why is this the case? Can't see how VS would or would not interfere with this definition.

jkoritzinsky replied:
This is outdated from a previous revision. I need to go remove this.

jaredpar commented on src/coreclr/src/System.Private.CoreLib/System.Private.CoreLib.csproj:
```diff
@@ -13,7 +13,7 @@
 
     <!-- Force System.Private.CoreLib.dll into a special IL output directory -->
-    <OutputPath>$(BinDir)IL/</OutputPath>
+    <OutputPath>$(BinDir)/IL/</OutputPath>
```
The rest of this project file seems to prefer \ over /

jkoritzinsky replied:
It looks like it's split across items and properties. Items use `\` whereas properties in this file tend to use `/`.

jaredpar commented on src/coreclr/src/System.Private.CoreLib/System.Private.CoreLib.csproj:
```diff
@@ -458,4 +436,56 @@
   <Import Project="$(RepositoryEngineeringDir)codeOptimization.targets" />
+
+  <!-- Setup eventing file generation -->
+  <Target Name="FindPythonWindows"
```
It feels a bit off to add a bunch of targets to the CoreCLR  build to discover the python executable.

Did you consider putting them in a more central location and then using a property, like `<RequiresPython>true</RequiresPython>` to allow projects to opt into using it? That would make it more generally useful and clean up this project file.

jkoritzinsky replied:
I'll go move them up so more projects can opt-in.

jkoritzinsky replied:
Now any target can depend on the `FindPython` target and python will be available for their scripts to use.

lpereira commented on eng/native/init-os-and-arch.sh:
```diff
@@ -0,0 +1,76 @@
+        echo "Unsupported OS $OSName detected, configuring as if for Linux"
+        os=Linux
+        ;;
+esac
```
Nitpicky suggestion:

```sh
case "$OSName" in
FreeBSD|Linux|NetBSD|OpenBSD|SunOS)
    os=$OSName ;;
Darwin)
	os=OSX ;;
*)
    echo "Unsupported OS $OSName detected, configuring as if for Linux"
    os=Linux ;;
esac
```

lpereira commented on eng/native/init-os-and-arch.sh:
```diff
@@ -0,0 +1,76 @@
+        fi
+        ;;
+
+    i686)
```
Might want to case on `i[3-6]86` here instead, to be thorough.  (Even though machines older than i686 are *really* unlikely these days.)

lpereira commented on eng/native/init-os-and-arch.sh:
```diff
@@ -0,0 +1,76 @@
+
+    amd64)
+        arch=x64
+        ;;
```
Maybe  case on `amd64|x86_64)` and get rid of the case below?

jkoritzinsky committed:
Remove dead MSBuild code.

jkoritzinsky committed:
Factor out python locating targets to make them available to the whole repo.

jkoritzinsky committed:
PR Feedback/simplification.

jkoritzinsky committed:
Merge branch 'master' of https://github.com/dotnet/runtime into hoist-S.P.CL

fadimounir commented:
>  can you test out a crossgen2 package from this PR to make sure it lays everything out correctly for the SDK to consume it?

@jkoritzinsky where can I find the package? in which artifact is it?

jkoritzinsky commented:
It looks like we don't publish the packages as artifacts in PR builds. Looks like we'll have to grab a local build. I can put one together for you and send it over.

jkoritzinsky committed:
Merge commit '9a8bce7' into hoist-S.P.CL

jkoritzinsky committed:
Create a self-contained deps.json for crossgen2 and add it to the package.

jkoritzinsky committed:
Merge branch 'master' of https://github.com/dotnet/runtime into hoist-S.P.CL

jkoritzinsky commented:
I've confirmed offline with Fadi that the package produced by the most recent commit in this PR works for the Crossgen2 SDK work.

jkoritzinsky committed:
Pass runtimeconfiguration as configuration for crossgen2 deps file generation.

jkoritzinsky committed:
Try specifying rids directly in crossgen2 project to get the restore correct for the live-build pipeline.

jkoritzinsky committed:
Run crossgen2 via corerun in R2R.SuperILC.

am11 commented on src/coreclr/src/tools/crossgen2/crossgen2/crossgen2.csproj:
```diff
@@ -7,16 +7,12 @@
+    <OutputPath>$(BinDir)/crossgen2</OutputPath>
     <GenerateRuntimeConfigurationFiles>false</GenerateRuntimeConfigurationFiles>
     <EnableDefaultEmbeddedResourceItems>false</EnableDefaultEmbeddedResourceItems>
+    <RuntimeIdentifiers>linux-x64;linux-musl-x64;win-x64</RuntimeIdentifiers>
```
Is it missing the macOS RID? Would it be possible to include all supported RID from the repos RID graph file here and in readytorun's csproj?

The reason why I am asking is; while working on Android (#32800), I found that this csproj and readytorun csproj were failing to build because of the missing RID. If there is no dependency that is preventing to use all-rids graph in live-live build scenario, please consider adding those. :)

jkoritzinsky replied:
This matches the rids specified in the `src/installer/pkg/projects/netcoreapp/pkg/crossgenRIDs.props` file to enable the live-live pipeline to correctly generate a project.assets.json that can be used to create a self-contained deps.json for the crossgen2 package. Without this, the live-build pipeline was failing. I couldn't repro the failure locally, but this logically fixed the failure.

am11 replied:
Ah, I think one way is to Import that prop file at the beginning (after the `<Project ...>` line).

I was actually referring to all supported platform RIDs in: https://github.com/dotnet/runtime/blob/bd98efe/src/libraries/pkg/Microsoft.NETCore.Platforms/runtime.json.  If there is no technical blocker, could crossgen2 and readytorun csprojs use that list without whitelisting just a handful of RIDs, like rest/most of the csprojs in the runtime repo?

jkoritzinsky replied:
I'll see if I can finagle it to have a shared list.

Crossgen2 has a native dependency (the JIT), so it'd have to be updated to be bundled with the correct JIT, which it currently only does for Windows and Linux. It'd probably get macOS x64 correct, but it definitely isn't set up to set up cross-builds correctly.

I believe the list above is the supported list of RIDs for the .NET 5 timeframe anyway.

jkoritzinsky replied:
What was the particular failure you were seeing in #32800 for these projects? If it was the self-contained publish, that wouldn't be an issue here since the way the project is restored it would still work if we're building for a platform whose RID isn't in the list.

jkoritzinsky replied:
The more I think about it, I don't think we can centralize the list of RIDs until the crossgen2 team decides how they're going to support the distros we cross-build for. For this PR I'll keep the two lists separate unless someone from that team tells me otherwise.

am11 replied:
> What was the particular failure you were seeing

The error was emerging from `src/coreclr/build.sh`, where it calls `dotnet publish` on `crossgen2.csproj` (after compiling native components and building managed components successfully):
```
 error : NETSDK1056: Project is targeting runtime 'android.28-arm64' but did not resolve any runtime-specific packages. This runtime may not be supported by the target framework.
```

maybe we can skip building this project to get full E2E clean build on other platforms. (I haven't tested the behavior with your branch, perhaps it is already like that with the deprecation of `coreclr/build.sh`)

jkoritzinsky replied:
Sounds like it might be a publish-only issue. The new scripts don’t run dotnet publish (which is why this specific line I added needed to be added). I’m pretty sure that won’t be a problem, although it’d be good to check it out if you have a chance.

am11 replied:
I tested the Android changes by rebasing against your branch. CoreCLR build succeeded (native and managed components), no more errors about crossgen2. 👍  Libraries build continues to succeed (native and managed components).

However, in installer build, it has progressed to fully build native (with correct RID) and managed components (with wrong RID), thereby packaging step fails because of the wrong RID was picked up during the ngen and perfmap generation steps:

```sh
$ ROOTFS_DIR=$(realpath .tools/android-rootfs/android-ndk-r21/sysroot) ./build.sh --cross --arch arm64 --subsetCategory installer
__DistroRid: android.28-arm64
__RuntimeId: android.28-arm64
  Tool 'coverlet.console' (version '1.7.0') was restored. Available commands: coverlet
  Tool 'dotnet-reportgenerator-globaltool' (version '4.4.2') was restored. Available commands: reportgenerator
  Tools restore was successful.
  Restore completed in 85.39 ms for /datadrive/home/.nuget/packages/microsoft.dotnet.arcade.sdk/5.0.0-beta.20124.3/tools/Tools.proj.
[snip]
  __DistroRid: android.28-arm64  <--- these guys show up twice with this branch?
  __RuntimeId: android.28-arm64
[snip]
  Commencing build of "installer component" for Linux.arm64.Debug in /datadrive/projects/runtime4/artifacts/obj/android.28-arm64.Debug
[snip]
 [100%] Linking CXX executable nativehost
  [100%] Built target nativehost
  Install the project...
  -- Install configuration: "DEBUG"
  -- Installing: /datadrive/projects/runtime4/artifacts/bin/android.28-arm64.Debug/corehost/apphost
  -- Installing: /datadrive/projects/runtime4/artifacts/bin/android.28-arm64.Debug/corehost/dotnet
[snip]   <!--  interesting part: note the RID in path a line above and below  -!>
  Native image /datadrive/projects/runtime4/artifacts/bin/linux-arm64.Debug/crossgen/netcoreapp/runtimes/linux-arm64/lib/netcoreapp5.0/Microsoft.CSharp.dll generated successfully.
  Native image /datadrive/projects/runtime4/artifacts/bin/linux-arm64.Debug/crossgen/netcoreapp/runtimes/linux-arm64/lib/netcoreapp5.0/Microsoft.VisualBasic.Core.dll generated successfully.
[snip] <!- build continues with wrong RID -!>
  Successfully generated perfmap for native assembly '/datadrive/projects/runtime4/artifacts/bin/linux-arm64.Debug/crossgen/netcoreapp/runtimes/linux-arm64/lib/netcoreapp5.0/Microsoft.CSharp.dll'.
  Successfully generated perfmap for native assembly '/datadrive/projects/runtime4/artifacts/bin/linux-arm64.Debug/crossgen/netcoreapp/runtimes/linux-arm64/lib/netcoreapp5.0/Microsoft.VisualBasic.Core.dll'.
[snip] <!- packaging step fails because RID of native (and manged) components of coreclr, libraries are different than what installer's ngen step picks up -!>
  Microsoft.NETCore.DotNetAppHost -> /datadrive/projects/runtime4/artifacts/packages/Debug/specs/Microsoft.NETCore.DotNetAppHost.nuspec
/datadrive/home/.nuget/packages/microsoft.dotnet.build.tasks.packaging/5.0.0-beta.20124.3/build/Packaging.targets(1260,5): error : Error when creating nuget lib package from /datadrive/projects/runtime4/artifacts/packages/Debug/specs/runtime.linux-arm64.Microsoft.NETCore.DotNetAppHost.nuspec. System.IO.DirectoryNotFoundException: Could not find a part of the path '/datadrive/projects/runtime4/artifacts/bin/linux-arm64.Debug/corehost'. [/datadrive/projects/runtime4/src/installer/pkg/projects/Microsoft.NETCore.DotNetAppHost/Microsoft.NETCore.DotNetAppHost.pkgproj]
/datadrive/home/.nuget/packages/microsoft.dotnet.build.tasks.packaging/5.0.0-beta.20124.3/build/Packaging.targets(1260,5): error :    at System.IO.Enumeration.FileSystemEnumerator`1.CreateDirectoryHandle(String path, Boolean ignoreNotFound) [/datadrive/projects/runtime4/src/installer/pkg/projects/Microsoft.NETCore.DotNetAppHost/Microsoft.NETCore.DotNetAppHost.pkgproj]
/datadrive/home/.nuget/packages/microsoft.dotnet.build.tasks.packaging/5.0.0-beta.20124.3/build/Packaging.targets(1260,5): error :    at System.IO.Enumeration.FileSystemEnumerator`1.Init() [/datadrive/projects/runtime4/src/installer/pkg/projects/Microsoft.NETCore.DotNetAppHost/Microsoft.NETCore.DotNetAppHost.pkgproj]
```

jkoritzinsky replied:
I'll file a new issue on Monday to track rid unification throughout the repo.

am11 replied:
Thanks!
A nit, this segment:
```
__DistroRid: android.28-arm64
__RuntimeId: android.28-arm64
```
appears twice in build out with PR changes.

jkoritzinsky replied:
That's expected currently since we run `init_distro_rid` twice, once in the root script and once in the coreclr-specific script. It's something we can clean up after this PR when we consolidate the RID discovery.

jkoritzinsky replied:
Filed #33069 to track unifying the RID discovery methods

jkoritzinsky committed:
Merge branch 'master' of https://github.com/dotnet/runtime into hoist-S.P.CL

jkoritzinsky committed:
Merge branch 'master' of https://github.com/dotnet/runtime into hoist-S.P.CL

jkoritzinsky committed:
Add binlogs to build-runtime.cmd

jkoritzinsky committed:
Combine non build-runtime steps into a single step so we get one binlog that doesn't get overwritten.

jkoritzinsky committed:
Add reference to issue describing new workflow.

jkoritzinsky commented:
@dotnet/runtime-infrastructure can I get another review pass on this PR?

jaredpar commented:
Reviewing

jaredpar commented:
Before merging we need to do an official signed build of this change. Too much of the core build logic is changing. Need to validate ahead of time it doesn't break official builds.

jaredpar commented on eng/python.targets:
```diff
@@ -0,0 +1,26 @@
+    <PropertyGroup>
+      <_PythonLocationScript>-c "import sys; sys.stdout.write(sys.executable)"</_PythonLocationScript>
+    </PropertyGroup>
+    <Exec Command="py -3 $(_PythonLocationScript) || py -2 $(_PythonLocationScript) || python3 $(_PythonLocationScript) || python2 $(_PythonLocationScript) || python $(_PythonLocationScript)"
```
Our build requires python 3 so why are we using `py -2` and `python2`? That's basically allowing the build to proceed when we know it's going to fail. Better to error early and push developers to supported toolsets.

jkoritzinsky replied:
I initially just pulled this from the old CoreCLR build script. I'll update it to only grab python3 installations.

jkoritzinsky commented on eng/Subsets.props:
```diff
@@ -81,7 +81,11 @@
 
     <!-- CoreClr -->
-    <SubsetName Include="All" Category="CoreClr" Description="The .NET runtime." />
+    <SubsetName Include="Runtime" Category="CoreClr" Description="The CoreCLR .NET runtime." />
```
This is the Runtime subset within the CoreCLR subset category. Mono could change their build to have a Runtime subset and then running `./build -subset runtime` would build both native runtimes.

jkoritzinsky commented on eng/Subsets.props:
```diff
@@ -61,7 +61,7 @@
     <DefaultLibrariesSubsets>all</DefaultLibrariesSubsets>
-    <DefaultCoreClrSubsets>all</DefaultCoreClrSubsets>
+    <DefaultCoreClrSubsets>runtime-corelib-nativecorelib-tools-packages</DefaultCoreClrSubsets>
     <DefaultMonoSubsets>all</DefaultMonoSubsets>
```
Mono's default doesn't include CoreCLR. Mono's default includes the "all" subset in the Mono subset category. Same way that Libraries' default includes the "all" subset in the Libraries subset category.

jaredpar replied:
Seems confusing to name it "all" then not have all the items in it. Is there some subtle, or less subtle, context I'm missing here?

jkoritzinsky replied:
The “all” refers to “all projects that are in this subset category” not “all projects in dotnet/runtime”.

jkoritzinsky committed:
Quote python path

jkoritzinsky committed:
Set arch default.

jkoritzinsky commented on eng/Subsets.props:
```diff
@@ -106,6 +110,18 @@
       <AdditionalProperties>Configuration=$(CoreCLRConfiguration)</AdditionalProperties>
     </CoreClrProjectToBuild>
+    <CoreClrCoreLibProjectToBuild>
+      <AdditionalProperties>Configuration=$(CoreCLRConfiguration)</AdditionalProperties>
```
This is done to match the above lines for `CoreClrProjectToBuild` to be consistent.

hoyosjs commented on src/coreclr/build-runtime.cmd:
```diff
@@ -0,0 +1,719 @@
+REM It is convenient to have your Nuget search path include the location where the build
+REM will place packages.  However nuget used during the build will fail if that directory
+REM does not exist.   Avoid this in at least one case by aggressively creating the directory.
+if not exist "%__BinDir%\.nuget\pkg"           md "%__BinDir%\.nuget\pkg"
```
where are our built packages used in native bits?

hoyosjs reviewed (approved):
Except the little breaking comments

hoyosjs commented:
> > To help make this more visible, I'm thinking about renaming the build.cmd/sh scripts to be build-native.cmd/sh and splitting out the crossgen script to be more explicit.
> 
> This renaming is a good idea that I strongly suggest. Instead of completely removing the `build.cmd/sh` scripts I would instead leave them intact for some amount of time (2 - 3 weeks) and have them print out the new workflow (i.e. point to a readme/URL/PR). The breadcrumb of what to do next would be very helpful.

Another thing: given the verbosity of these scripts, can we put the message as the end of the build?

jkoritzinsky commented on src/coreclr/src/scripts/check-definitions.py:
```diff
@@ -1,163 +0,0 @@
-#!/usr/bin/env python
```
The definition consistency checking never really worked. It’s had warnings for months and no one could see them since we moved it into Arcade. It also doesn’t work with CMake Generator Expressions. @janvorli said it was fine if we removed it.

hoyosjs commented on src/coreclr/build-runtime.cmd:
```diff
@@ -0,0 +1,719 @@
+set "__SourceDir=%__ProjectDir%\src"
+set "__RootBinDir=%__RepoRootDir%\artifacts"
+set "__LogsDir=%__RootBinDir%\log"
+set "__MsbuildDebugLogsDir=%__LogsDir%\MsbuildDebugLogs"
```
```suggestion
set "__LogsDir=%__RootBinDir%\log\!__BuildType!"
set "__MsbuildDebugLogsDir=%__LogsDir%\MsbuildDebugLogs"
```
Although it needs to be further down around __CrossCompIntermediatesDir

jkoritzinsky commented on eng/Subsets.props:
```diff
@@ -81,7 +81,11 @@
-    <SubsetName Include="All" Category="CoreClr" Description="The .NET runtime." />
+    <SubsetName Include="Runtime" Category="CoreClr" Description="The CoreCLR .NET runtime." />
+    <SubsetName Include="CoreLib" Category="CoreClr" Description="The managed System.Private.CoreLib library for CoreCLR." />
+    <SubsetName Include="NativeCoreLib" Category="CoreClr" Description="Run crossgen on System.Private.CoreLib library for CoreCLR." />
```
The parameter in the current scripts to skip it is skipnativecorlib I believe, so I tried to match the original naming design.

jkoritzinsky commented on eng/build.ps1:
```diff
@@ -146,11 +148,13 @@ foreach ($argument in $PSBoundParameters.Keys)
   }
 }
 
+$env:__DistroRid="win-$arch"
```
It was only getting set during the CoreCLR build in the old scripts.

jkoritzinsky committed:
PR Feedback and clean up crossgen-corelib's argument list.

jkoritzinsky committed:
Fix subset name matching to work when one subset name is a substring of another.

jkoritzinsky commented on eng/pipelines/coreclr/templates/build-job.yml:
```diff
@@ -84,9 +84,13 @@ jobs:
+    - name: officialBuildIdArcadeArg
+      value: ''
     - ${{ if and(eq(variables['System.TeamProject'], 'internal'), ne(variables['Build.Reason'], 'PullRequest')) }}:
       - name: officialBuildIdArg
```
Yep. And it looks like I might be able to simplify this based on that. I'll go clean this up.

ViktorHofer commented on Directory.Build.targets:
```diff
@@ -9,6 +9,7 @@
   </PropertyGroup>
 
   <Import Project="$(RepositoryEngineeringDir)/liveBuilds.targets" />
+  <Import Project="$(RepositoryEngineeringDir)/python.targets" />
```
nit: remove the slash as `RepositoryEngineeringDir` already has it in it: https://github.com/dotnet/runtime/blob/master/Directory.Build.props#L24 and https://github.com/dotnet/arcade/blob/e0112fe2f1fd7f6db08b228db98f22225f82267b/src/Microsoft.DotNet.Arcade.Sdk/tools/RepoLayout.props#L44

ViktorHofer commented on eng/build.ps1:
```diff
@@ -146,11 +146,13 @@ foreach ($argument in $PSBoundParameters.Keys)
     "os"                   { $arguments += " /p:OSGroup=$($PSBoundParameters[$argument])" }
     "allconfigurations"    { $arguments += " /p:BuildAllConfigurations=true" }
-    "arch"                 { $arguments += " /p:ArchGroup=$($PSBoundParameters[$argument]) /p:TargetArchitecture=$($PSBoundParameters[$argument])" }
+    "arch"                 { $arch = $PSBoundParameters[$argument]; $arguments += " /p:ArchGroup=$($PSBoundParameters[$argument]) /p:TargetArchitecture=$($PSBoundParameters[$argument])" }
```
nit: use the arch variable in the subsequent statements in the same line.

ViktorHofer commented on eng/run-test.sh:
```diff
@@ -65,60 +65,12 @@ trap handle_ctrl_c INT
 
 # Location parameters
-# OS/Configuration defaults
+# OS/ConfigurationGroup defaults
```
Even though this is just a comment I would revert this back to `Configuration`. We don't use `ConfigurationGroup` anymore.

jkoritzinsky committed:
Pass the official build id as an msbuild arg at all times.
Additional cleanup for crossgen-corelib step.

jkoritzinsky committed:
PR Feedback.

jkoritzinsky committed:
Fix default subset list construction.

jkoritzinsky committed:
Add deprecation warnings at beginning and end of build.cmd and build.sh.

jkoritzinsky commented:
There's no single exit point for src/coreclr/build.sh, so I'm just going to put the warning at the beginning and at the "success" end.

jkoritzinsky committed:
Setup dirs in crossgen-corelib.sh

jkoritzinsky commented:
Official build run at: https://dev.azure.com/dnceng/internal/_build/results?buildId=543897&view=results

jkoritzinsky committed:
Move enforcepgo argument to before the official build arg so that it gets consumed correctly.

jkoritzinsky committed:
Merge branch 'master' of https://github.com/dotnet/runtime into hoist-S.P.CL

jkoritzinsky committed:
Add cross-os DAC builds as subsets.
Clean up build-runtime parameters and passing the official build ID from runtime.proj.

jkoritzinsky commented:
Yep. We're passing it as an "unprocessed build arg", which gets passed along to the MSBuild invocation for GenerateNativeVersionFile. If you check the official build I linked to, the outputs of the CoreCLR native build have the correct version info.

jkoritzinsky committed:
Update script usage docs.

jkoritzinsky committed:
Skip the cross-os dac on x86.

jkoritzinsky committed:
Merge branch 'master' of https://github.com/dotnet/runtime into hoist-S.P.CL

jkoritzinsky commented:
Only failure is due to https://github.com/dotnet/runtime/issues/32764

Anyone else have any feedback? If not, I'm going to merge this at EOD.

jkoritzinsky committed:
Build windows->linux cross dac on PR and CI.

jkoritzinsky committed:
Don't build the cross-OS linux dac on Windows x86.
